zerver/actions/message_send.py

```diff
@@ -731,6 +743,15 @@ def create_user_messages(
             flags |= UserMessage.flags.mentioned
         if user_profile_id in ids_with_alert_words:
             flags |= UserMessage.flags.has_alert_word
+        if (
+            not rendering_result.mentions_stream_wildcard
```

Comment:
This `not` is a noop, because that flag will already be there, right? Might be cleaner to not have it.

Reply:
Suppose a message has both `stream` and `topic` wildcard mentions.

We already set the `wildcard_mentioned` flag for all the `um_eligible_user_ids` earlier in this function due to the stream wildcard mention. 
```python
    base_flags = 0
    if rendering_result.mentions_stream_wildcard:
        base_flags |= UserMessage.flags.wildcard_mentioned
```
So using this `not`, we can skip the following operation in the `if` conditional
```python
user_profile_id in topic_wildcard_mention_user_ids.union(
    topic_wildcard_mention_in_followed_topic_user_ids
)
```

In simple words, if the message has both the stream and topic wildcard mention we can skip the checks related to topic wildcard (using this not) as we already have set the `wildcard_mentioned` flag for everyone. No need to check and set it again for individual users.

Reply:
Suppose a message has both `stream` and `topic` wildcard mentions.

We already set the `wildcard_mentioned` flag for all the `um_eligible_user_ids` earlier in this function due to the stream wildcard mention. 
```python
    base_flags = 0
    if rendering_result.mentions_stream_wildcard:
        base_flags |= UserMessage.flags.wildcard_mentioned
```

So using this `not`, we can skip the following operation in the if conditional:

```python
user_profile_id in topic_wildcard_mention_user_ids.union(
   topic_wildcard_mention_in_followed_topic_user_ids
)
```

In simple words, if the message has both the stream and topic wildcard mention, we can skip the checks related to topic wildcard (using this `not`), as we already have set the `wildcard_mentioned` flag for everyone. There is no need to check and set it again for individual users.

Reply:
One improvement I made:

Instead of performing the union operation within a loop, I extracted that as a separate variable named `all_topic_wildcard_mention_user_ids`

Reply:
> we can skip the checks related to topic wildcard (using this not)

That makes sense, but I'm not sure if this optimization would make that much of a difference, since all the data here is in-memory?

Reply:
Hm, right. Earlier the union operation was there in the if statement itself. Since I've extracted it out, we can skip this `not` check.